eth/db/header.py

```diff
@@ -146,46 +146,55 @@ def header_exists(self, block_hash: Hash32) -> bool:
         validate_word(block_hash, title="Block Hash")
         return block_hash in self.db
 
-    def persist_header(self,
-                       header: BlockHeader
-                       ) -> Tuple[Tuple[BlockHeader, ...], Tuple[BlockHeader, ...]]:
+    def persist_header_chain(self,
+                             *headers: BlockHeader
+                             ) -> Tuple[Tuple[BlockHeader, ...], Tuple[BlockHeader, ...]]:
         """
-        :returns: iterable of headers newly on the canonical chain
+        Return two iterable of headers, the first containing the new canonical headers,
+        the second containing the old canonical headers
         """
-        if header.parent_hash != GENESIS_PARENT_HASH:
-            try:
-                self.get_block_header_by_hash(header.parent_hash)
-            except HeaderNotFound:
+
+        previous = None
+
+        for header in headers:
+            is_genesis = header.parent_hash == GENESIS_PARENT_HASH
+            if not is_genesis and not self.header_exists(header.parent_hash):
                 raise ParentNotFound(
                     "Cannot persist block header ({}) with unknown parent ({})".format(
                         encode_hex(header.hash), encode_hex(header.parent_hash)))
 
-        self.db.set(
-            header.hash,
-            rlp.encode(header),
-        )
+            self.db.set(
+                header.hash,
+                rlp.encode(header),
+            )
 
-        if header.parent_hash == GENESIS_PARENT_HASH:
-            score = header.difficulty
-        else:
-            score = self.get_score(header.parent_hash) + header.difficulty
+            if is_genesis:
+                score = header.difficulty
+            elif previous is not None and previous[0].hash is header.parent_hash:
```

Comment:
If we add the validation above, we no longer have to confirm that the parent hash matches here. In fact, I think `previous` can just become `parent_score`, and generated before the loop, which removes this whole if/elif block.

Reply:
Yep, that was build with the goal to support non-contiguous chains as well https://github.com/ethereum/py-evm/pull/1192#discussion_r210939381

Reply:
I'd say just pick the first option he suggested, and save non-contiguous support for another PR.